### ISSUE TITLE
Distill teacher models where the vocab size of teacher and student is different

### ISSUE THREAD

**Issue by shaunakjoshi12** [huggingface/trl#3028]
I am trying to distill a Qwen2.5-7B-Instruct to Qwen2.5-5B-Instruct using a sample code 

```from datasets import Dataset
from trl import GKDConfig, GKDTrainer
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
)

NUM_DUMMY_SAMPLES = 100

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")

teacher_model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-7B-Instruct")

train_dataset = Dataset.from_dict(
    {
        "messages": [
            [
                {"role": "user", "content": "Hi, how are you?"},
                {"role": "assistant", "content": "I'm great thanks"},
            ]
        ]
        * NUM_DUMMY_SAMPLES
    }
)
eval_dataset = Dataset.from_dict(
    {
        "messages": [
            [
                {"role": "user", "content": "What colour is the sky?"},
                {"role": "assistant", "content": "The sky is blue"},
            ]
        ]
        * NUM_DUMMY_SAMPLES
    }
)

training_args = GKDConfig(output_dir="gkd-model", per_device_train_batch_size=1)
trainer = GKDTrainer(
    model=model,
    teacher_model=teacher_model,
    args=training_args,
    processing_class=tokenizer,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
)
trainer.train()```

But this gives me an error because their vocab sizes are different (so might be their tokenizers). Is there a workaround for these kind of situations? How are such cases handled?

**Comment by AamodThakur** [comment]
Hi, @shaunakjoshi12, I believe generic KD does not work with different vocabulary. This is a research problem where multiple papers are being published with a method to do knowledge transfer to models with different vocabularies.

**Comment by qgallouedec** [comment]
Hey @shaunakjoshi12, sorry for the very late reply we're catching up on the backlog.

The crash comes from GKD's loss: it's a token-level KL divergence between teacher and student logits, see https://github.com/huggingface/trl/blob/6f914487dbec1e61c7c297f936cd4f9c0005e39a/trl/experimental/gkd/gkd_trainer.py#L266-L269
so it needs both models to share a vocabulary. In your case Qwen2.5-0.5B and Qwen2.5-7B actually use the *same* tokenizer (max token id 151664 in both), but their config `vocab_size` differs (151936 vs 152064 — just padded embedding rows), so the logits can't be combined. For genuinely different vocab sizes / tokenizers, TRL now ships an experimental cross-tokenizer trainer, `GOLDTrainer` (`from trl.experimental.gold import GOLDConfig, GOLDTrainer`), which implements Universal Logit Distillation and is built for exactly this case — see #4562 for the full discussion. Closing as a duplicate of that thread, but happy to reopen if GOLD doesn't cover your setup.
